config/config.go

```diff
@@ -75,29 +69,141 @@ var (
+	Email string `yaml:"email"`
+
 	// Token is the session token of the authenticated user.
-	Token string `yaml:"token,omitempty"`
+	Token string `yaml:"token"`
+}
 
-	// Updated is the time when the config has last been written.
-	Updated string `yaml:"updated,omitempty"`
+// StoreEndpointAuth adds an endpoint to the configStruct.Endpoints field
+// (if not yet there). This should only be done after successful authentication.
+func (c *configStruct) StoreEndpointAuth(endpointURL string, email string, token string) error {
+	ep := normalizeEndpoint(endpointURL)
+
+	if email == "" || token == "" {
+		return microerror.Mask(credentialsRequiredError)
+	}
+
+	if c.Endpoints == nil {
+		c.Endpoints = map[string]*endpointConfigStruct{}
+	}
+
+	c.Endpoints[ep] = &endpointConfigStruct{
+		Email: email,
+		Token: token,
+	}
+
+	WriteToFile()
+
+	return nil
 }
 
-// init sets up viper and sets defaults
-func init() {
-	viper.SetConfigType(ConfigFileType)
-	viper.SetConfigName(ConfigFileName)
-	viper.SetEnvPrefix(ProgramName)
-	viper.SetTypeByDefaultValue(true)
+// SelectEndpoint makes the given endpoint URL the selected one
+func (c *configStruct) SelectEndpoint(endpointURL string) error {
+	ep := normalizeEndpoint(endpointURL)
+	if _, ok := c.Endpoints[ep]; !ok {
+		return microerror.Mask(endpointNotDefinedError)
+	}
+
+	c.SelectedEndpoint = ep
+	c.Token = c.Endpoints[ep].Token
+	c.Email = c.Endpoints[ep].Email
+
+	WriteToFile()
+
+	return nil
+}
+
+// SelectedEndpoint returns the selected endpoint URL.
+// If the argument overridingEndpointURL is not empty, this will
+// be used as the returned endpoint URL.
+// Errors are only printed to inform users, but not returned, to simplify
+// usage of this function.
+func (c *configStruct) ChooseEndpoint(overridingEndpointURL string) string {
+	if overridingEndpointURL != "" {
+		ep := normalizeEndpoint(overridingEndpointURL)
+		return ep
+	}
+
+	envEndpoint := os.Getenv("GSCTL_ENDPOINT")
+	if envEndpoint != "" {
+		ep := normalizeEndpoint(envEndpoint)
+		return ep
+	}
+
+	return c.SelectedEndpoint
+}
+
+// ChooseToken chooses a token to use, according to a rule set.
+// - If the given token is not empty, we use (return) that
+// - If the given token is empty and we have an auth token for the given
+//   endpoint, we return that
+// - otherwise we return an empty string
+func (c *configStruct) ChooseToken(endpoint, overridingToken string) string {
+	ep := normalizeEndpoint(endpoint)
+
+	if overridingToken != "" {
+		return overridingToken
+	}
 
+	if endpointStruct, ok := c.Endpoints[ep]; ok {
+		if endpointStruct != nil && endpointStruct.Token != "" {
+			return endpointStruct.Token
+		}
+	}
+
+	return ""
+}
+
+// Logout removes the token value from the selected endpoint.
+func (c *configStruct) Logout(endpointURL string) {
+	ep := normalizeEndpoint(endpointURL)
+
+	if ep == c.SelectedEndpoint {
+		c.Token = ""
+	}
+
+	if element, ok := c.Endpoints[ep]; ok {
+		element.Token = ""
+	}
```

Comment:
Why not `delete(c.Endpoints, ep)`?

Reply:
I want to keep the endpoint entry with the email address stored (`element.Email`), to simplify future logins.